pyvisa/resources/messagebased.py

```diff
@@ -469,17 +469,44 @@ def read_binary_values(self, datatype='f', is_big_endian=False,
         :rtype: type(container)
 
         """
-        block = self._read_raw()
+        block = b''
         expected_length = 0
 
         if header_fmt == 'ieee':
-            offset, data_length = util.parse_ieee_block_header(block)
+            hash = self.read_bytes(1)                   # Hash '#'
+            digits = self.read_bytes(1)                 # Digits
+            block_size = self.read_bytes(int(digits))   # Block size
+
+            block = block + hash
+            block = block + digits
+            block = block + block_size
+
+            block = block + self.read_bytes(int(block_size))
+
+            offset = 2 + int(digits)
+            data_length = int(block_size)
+
             expected_length = offset + data_length
+
         elif header_fmt == 'hp':
-            offset, data_length = util.parse_hp_block_header(block,
-                                                             is_big_endian)
+
+            hash = self.read_bytes(2)                   # Hash '#A'
+            block_size = self.read_bytes(4)             # Block size
```

Comment:
Once again it makes sense to read all 6 bytes at once and then simply use the `parse_hp_block_header` utility to avoid code duplication.